### PR TITLE
Removing basepath/hostname from admin site settings

### PR DESCRIPTION
**Pull request by nupplaphil**
Replacing #6876 
Prerequisite of #6815 
Part of solving #6720 

- Removing the `$basepath` option from the admin panel (as said at #6720 , it is a base setting which shouldn't be able to change easily)
- Removing the `$hostname` option form the admin panel (because renaming the hostname leads to unexpected redirects -> f.e. the current users don't automatically move to new hostname)
- Refactoring `$this->basepath` to `$this->getBasePath()` as prerequisite for #6815 
- Refactoring `Core\Config::get()` to `$this->config->get()` as prerequisite for #6815 

ToDos:
- [x] Create relocation function for hostname in file
- [x] Create migration script for hostname and basepath from DB to file
- [x] Add a failure handling for permission/write problems
- [x] Add a "all in one" writer for config file changes

### CODE REVIEW

**Comment by nupplaphil** [comment]
- Tested with dbupdate function (`update_1305()`)
- Tested relocation with hostname
- Tested with `config`, `ini`, `.htconfig`
- Added some unittests to avoid regressions

If something goes wrong during the update (wrong file permission):
1) Inside Log
```console
2019-03-24 21:41:38 worker [INFO]: Update finished. {"from":1304,"to":1305} - {"file":"Update.php","line":137,"function":"run","uid":"56564f","process_id":18686}
2019-03-24 21:41:38 worker [INFO]: Update function start. {"function":"update_1305"} - {"file":"Update.php","line":177,"function":"runUpdateFunction","uid":"56564f","process_id":18686}
2019-03-24 21:41:38 worker [INFO]: Difference in config found {"cat":"config","key":"hostname","file":null,"saved":"friendica.local"} - {"file":"Update.php","line":287,"function":"updateConfigEntry","uid":"56564f","process_id":18686}
2019-03-24 21:41:38 worker [ALERT]: Config entry update failed - maybe wrong permission? [] - {"file":"Update.php","line":253,"function":"saveConfigToFile","uid":"56564f","process_id":18686}
2019-03-24 21:41:38 worker [WARNING]: subject isn't set. {"type":16384} - {"file":"enotify.php","line":457,"function":"notification","uid":"56564f","process_id":18686}
2019-03-24 21:41:38 worker [INFO]: sending notification email [] - {"file":"enotify.php","line":581,"function":"notification","uid":"56564f","process_id":18686}
2019-03-24 21:41:38 worker [INFO]: Path "/tmp/itemcache" is not writable for user www-data [] - {"file":"System.php","line":328,"function":"isDirectoryUsable","uid":"56564f","process_id":18686}
2019-03-24 21:41:38 worker [INFO]: Path "/tmp/itemcache" is not writable for user www-data [] - {"file":"System.php","line":328,"function":"isDirectoryUsable","uid":"56564f","process_id":18686}
2019-03-24 21:41:38 worker [INFO]: header To: admin@friendica.local
Precedence: list
X-Friendica-Host: friendica.local
X-Friendica-Account: <@friendica.local>
X-Friendica-Platform: Friendica
X-Friendica-Version: 2019.06-dev
List-ID: <notification.friendica.local>
List-Archive: <http://friendica.local/notifications/system>
From: Friendica Social Network <noreply@friendica.local>
Reply-To: Friendica Social Network <noreply@friendica.local>
MIME-Version: 1.0
Content-Type: multipart/alternative; boundary="8-564726580-526190717=:97071" [] - {"file":"Emailer.php","line":102,"function":"send","uid":"56564f","process_id":18686}
2019-03-24 21:41:38 worker [INFO]: return value true [] - {"file":"Emailer.php","line":103,"function":"send","uid":"56564f","process_id":18686}
2019-03-24 21:41:38 worker [ALERT]: Database structure update FAILED. {"error":"Update 1305 failed. See error logs."} - {"file":"Update.php","line":348,"function":"updateFailed","uid":"56564f","process_id":18686}
2019-03-24 21:41:38 worker [ERROR]: Update function ERROR. {"function":"update_1305","retval":1} - {"file":"Update.php","line":198,"function":"runUpdateFunction","uid":"56564f","process_id":18686}
```

2) Admin page
Screenshot_2019-03-24 Friendica Social Network Admin

3) Console update
`root@friendica:/vagrant# php bin/console.php dbstructure update`

```console
PHP Warning:  fopen(/vagrant/config/local.config.php.tmp): failed to open stream: Protocol error in /vagrant/src/Util/Config/ConfigFileSaver.php on line 138
PHP Stack trace:
PHP   1. {main}() /vagrant/bin/console.php:0
PHP   2. Asika\SimpleConsole\Console->execute() /vagrant/bin/console.php:11
PHP   3. Friendica\Core\Console->doExecute() /vagrant/vendor/asika/simple-console/src/Console.php:108
PHP   4. Asika\SimpleConsole\Console->execute() /vagrant/src/Core/Console.php:106
PHP   5. Friendica\Core\Console\DatabaseStructure->doExecute() /vagrant/vendor/asika/simple-console/src/Console.php:108
PHP   6. Friendica\Core\Update::run() /vagrant/src/Core/Console/DatabaseStructure.php:74
PHP   7. Friendica\Core\Update::runUpdateFunction() /vagrant/src/Core/Update.php:140
PHP   8. update_1305() /vagrant/src/Core/Update.php:188
PHP   9. Friendica\Core\Update::saveConfigToFile() /vagrant/update.php:358
PHP  10. Friendica\Util\Config\ConfigFileSaver->saveToConfigFile() /vagrant/src/Core/Update.php:250
PHP  11. Friendica\Util\Config\ConfigFileSaver->openFile() /vagrant/src/Util/Config/ConfigFileSaver.php:77
PHP  12. fopen() /vagrant/src/Util/Config/ConfigFileSaver.php:138
```

4) Resetted permission and update worked

**Comment by nupplaphil** [comment]
Merging https://github.com/friendica/friendica-addons/pull/832 too.

File: src/Util/Config/ConfigFileSaver.php [hunk 1]
```diff
@@ -0,0 +1,341 @@
...
+
+	/**
+	 * Save all added configuration entries to the given config files
+	 * After updating the config entries, all configuration entries will be reseted
+	 *
+	 * @param string $name The name of the configuration file (default is empty, which means the default name each type)
+	 *
+	 * @return bool true, if at least one configuration file was successfully updated or nothing to do
+	 */
+	public function saveToConfigFile($name = '')
```
Comment (MrPetovan) [inline src/Util/Config/ConfigFileSaver.php]: Correct me if I'm wrong, but isn't this method able to automatically convert any install config files to the latest format?

You see where I'm going with this, right?

**Comment by MrPetovan** [comment]
After pulling this PR, my whole display broke because the `system.basepath` value is empty.

**Comment by MrPetovan** [comment]
This is my manual solution for now: https://friendica.mrpetovan.com/display/735a2029-185c-9847-b083-484188242854

**Comment by nupplaphil** [comment]
I always had a system.basepath in the db during the try.. I'll fix it soon

File: src/Util/Config/ConfigFileSaver.php (hunk 1 above)
Comment (nupplaphil) [reply src/Util/Config/ConfigFileSaver.php]: Ha, yes that's possible :-)

So we create an automatic upgrade path for legacy files. That sounds exciting

**Pull request merged**